psfws/gen_params.py

```diff
@@ -0,0 +1,385 @@
+        self.h = h[where_h0:where_end]
+        self.p = p_and_h[src]['p'][::-1][where_h0:where_end]
+
+        for k in ['u', 'v', 't', 'speed', 'phi']:
+            forecast[k] = [forecast[k].values[i][where_h0:where_end] 
+                           for i in range(self.N)]
+        self.data_fa = forecast
+
+        # ground layer ends at ~1km above telescope, save for later 
+        self.fa_start = np.searchsorted(self.h, self.h0 + 0.8)
+        
+    def get_measurements(self, pt):
+        """Get a matched set of measurements from datapoint with index ``pt``.
+
+        Parameters:
+            pt:     Pandas Timestamp of date/time for desired output.
+
+        Returns:
+            params: Dictionary of wind and temperature measurements, made of
+                    matched ground level and free atmosphere data, at Index 
+                    location ``pt``. Keys are [``u``, ``v``, ``speed``, ``phi``,
+                    ``t``, ``h``] (see class Attributes for details).
+        """
+        try:
+            gl = self.data_gl.loc[pt]
+            fa = self.data_fa.loc[pt]
+        except KeyError:
+            if type(pt) == str:
+                raise TypeError(f'pt type must be pd.Timestamp not str!')
+            if type(pt) == pd.Timestamp:
+                raise KeyError(f'{pt} not found in data index!')
+
+        direction = np.hstack([gl.at['phi'], fa.at['phi']])
+
+        return {'u': np.hstack([gl.at['u'], fa.at['u']]),
+                'v': np.hstack([gl.at['v'], fa.at['v']]),
+                'speed': np.hstack([gl.at['speed'], fa.at['speed']]),
+                't': np.hstack([gl.at['t'], fa.at['t']]),
+                'h': np.hstack([self.h0, self.h]),
+                'phi': utils.smooth_dir(direction)}
+
+    def _interpolate(self, p_dict, h_out, s=None):
+        """Get interpolations & derivatives of p_dict at new heights h_out."""
+        # check h_out values
+        if max(h_out) > 100:
+            # 100km would be an outrageously high altitude, probably unit error
+            raise ValueError('Units of h should be in km, not m.')
+        if min(h_out) < min(self.h) or max(h_out) > max(self.h):
+            raise ValueError(f"Can't interpolate below {min(self.h)}" + \
+                             f" or above {max(self.h)}.")
+
+        # if 'h' in the dictionary, use those heights
+        if 'h' in p_dict.keys():
+            h_in = p_dict['h']
+        # otherwise inputs are raw FA values so use self.h
+        else:
+            h_in = self.h
+
+        out = {}
+        # Note: multipying by 1000 (to m) for derivative units.
+        for k in ['u', 'v', 't']:
+            out[k], out[f'd{k}dz'] = utils.interpolate(h_in * 1000,
+                                                       p_dict[k],
+                                                       h_out * 1000,
+                                                       s=s)
+
+        # special case (h is always self.h here)
```

Comment:
Weather forecast models use sets of equipotential heights for outputting quantities -- ie the outputs are v(p) not v(h) -- so the set of pressures through the atmosphere is always the same. I used a simple model of h(p) to convert those to altitudes for more intuitive use. I've made both h and p class attributes, self.h and self.p. 
So p isn't in the parameter dict that gets passed to this interpolate function, since those will just be the time-dependent variables (wind info and temperature), and self.h is by definition the set of associated heights for self.p.